assets/app/scripts/services/data.js

```diff
@@ -199,15 +131,271 @@ angular.module('openshiftConsole')
+    if (!op && op !== false) {
+      return this._listOperationMap[key];
+    }
+    else {
+      this._listOperationMap[key] = op;
+    }
+  };
+
+  DataService.prototype._resourceVersion = function(type, context, rv) {
+    var key = this._uniqueKeyForTypeContext(type, context);
+    if (!rv) {
+      return this._resourceVersionMap[key];
+    }
+    else {
+      this._resourceVersionMap[key] = rv;
+    }
+  };
+
+  DataService.prototype._data = function(type, context, data) {
+    var key = this._uniqueKeyForTypeContext(type, context);
+    if (!data) {
+      return this._dataMap[key];
+    }
+    else {
+      this._dataMap[key] = new Data(data);
+    }
+  }; 
+
+  DataService.prototype._watchPollTimeouts = function(type, context, timeout) {
+    var key = this._uniqueKeyForTypeContext(type, context);
+    if (!timeout) {
+      return this._watchPollTimeoutsMap[key];
+    }
+    else {
+      this._watchPollTimeoutsMap[key] = timeout;
+    }
+  }; 
+
+  DataService.prototype._watchWebsockets = function(type, context, timeout) {
+    var key = this._uniqueKeyForTypeContext(type, context);
+    if (!timeout) {
+      return this._watchWebsocketsMap[key];
+    }
+    else {
+      this._watchWebsocketsMap[key] = timeout;
+    }
+  };     
+
+  DataService.prototype._uniqueKeyForTypeContext = function(type, context) {
+    // Note: when we start handling selecting multiple projects this
+    // will change to include all relevant scope
+    return type + context.projectName;
+  };
+
+  DataService.prototype._startListOp = function(type, context, opts) {
+    // mark the operation as in progress
+    this._listInFlight(type, context, true);
+
+    if (context.projectPromise && type !== "projects") {
+      context.projectPromise.done($.proxy(function(project) {
+        $.ajax({
+          url: this._urlForType(type, null, context, false, {namespace: project.metadata.namespace}),
+          success: $.proxy(this, "_listOpComplete", type, context, opts)
+        });
+      }, this));
+    }
+    else {
+      $.ajax({
+        url: this._urlForType(type, null, context),
+        success: $.proxy(this, "_listOpComplete", type, context, opts)
+      });
+    }
+  };
+
+  DataService.prototype._listOpComplete = function(type, context, opts, data) {
+    this._resourceVersion(type, context, data.resourceVersion || data.metadata.resourceVersion);
+    this._data(type, context, data.items);
+    this._listCallbacks(type, context).fire(this._data(type, context));
+    this._listCallbacks(type, context).empty();    
+    this._watchCallbacks(type, context).fire(this._data(type, context));
+
+    // mark list op as complete
+    this._listInFlight(type, context, false);
+
+    if (this._watchCallbacks(type, context).has()) {
+      if (opts && opts.poll) {
+        this._watchPollTimeouts(type, context, setTimeout($.proxy(this, "_startListOp", type, context, opts), opts.pollInterval || 5000));
```

Comment:
do we need to check if a list op is already in flight? we could get two competing with each other, one polling, one not...